Cleanup and clarify recurrences src / Fix #100

Datseris opened the pull request:
I am working on functionality that given a dynamical system it animates the action of `AttractorsViaRecurrences`. I will contribute this in a separate PR, but I am doing pretty good, here is where I am now: 

https://github.com/JuliaDynamics/Attractors.jl/assets/19669089/d4e856fe-85da-473d-b00b-10f8211c9740

During this journey, I had some problems with our source code. I admit, I find it surprising how difficult the source code is to read, given the success of the algorithm.

In this PR I do the following:

- Re-organize the source code more logically
- Re-named many variables and functions so that they better convey what they do. E.g., `visited_cell` was a poor name. It contained the (default) label that visited cells obtain. You'd think it would contain a cell (i.e., a cartesian index instead). A better name is `visited_cell_label`. Etc.
- Fix #100 
- Add many more comments to the source code to better explain things

Datseris commented:
oooof oh boy this is finally done. Took me quite a while to go through the source code and really clarify things as much as possible.

@awage or @kalel can you please review and leave an Approving review if you don't find anything wrong so that I can merge this in? This PR does not alter any functionality, it only re-wrote the source code to be more readable.

awage commented:
Hi! I will review the code today and leave some comments if any.

awage commented on src/mapping/recurrences/attractor_mapping_recurrences.jl:
```diff
@@ -0,0 +1,326 @@
+  limit we declare that the orbit diverged to infinity.
+* `mx_chk_safety = Int(1e6)`: A safety counter that is always increasing for
+  each initial condition. Once exceeded, the algorithm assigns `-1` and throws a warning.
+  This clause exists to stop the algorithm never haulting for innappropriate grids,
+  where a found attractor may intersect in the same cell with a new attractor the orbit
+  traces (which leads to infinite resetting of all counters).
```
```suggestion
This clause exists to stop the algorithm when it never halts due to inappropriate grids. For example it happens when the orbit of a newly found attractor intersect in the same cell of a previously found attractor  (which leads to infinite resetting of all counters) ```

awage replied:
```suggestion
  This clause exists to stop the algorithm never halting for inappropriate grids. It may happen 
  when a newly found attractor orbit intersects in the same cell of a previously found attractor (which leads to infinite resetting of all counters).
```

awage commented on src/mapping/recurrences/attractor_mapping_recurrences.jl:
```diff
@@ -0,0 +1,326 @@
+    prev_label::Int
+    safety_counter::Int
+    attractors::Dict{Int, StateSpaceSet{D, T}}
+    visited_list::Vector{CartesianIndex{D}}
```
`visited_list` can be renamed as `visited_cells` or `visited_cells_list`. Technically it is not a list, it is a stack.

awage commented on src/mapping/recurrences/finite_state_machine.jl:
```diff
@@ -0,0 +1,275 @@
+    bsn_nfo.state = :att_search
+end
+
+function update_fsm_state!(bsn_nfo, ic_label)
```
This function is very cryptic. I am sure there is a better way to express the transitions from state to state. I will try to improve this part.

awage commented on src/mapping/recurrences/attractor_mapping_recurrences.jl:
```diff
@@ -0,0 +1,326 @@
+#####################################################################################
+# Definition of `BasinInfo` and initialization
+#####################################################################################
+abstract type Grid end
```
This type should be in grids.jl, there is a reason to declare this here?

awage commented on src/mapping/recurrences/grids.jl:
```diff
@@ -0,0 +1,169 @@
+
+Given a coarse `grid` tesselating the state space, construct a `SubdivisionBasedGrid`
+based on the given level array `lvl_array`.
+The level array is has non-negative integer values.
```
```suggestion
The level array has non-negative integer values.
```

awage commented on src/mapping/recurrences/grids.jl:
```diff
@@ -0,0 +1,169 @@
+"""
+    SubdivisionBasedGrid(grid::NTuple{D, <:AbstractRange}, lvl_array::Array{Int, D})
+
+Given a coarse `grid` tesselating the state space, construct a `SubdivisionBasedGrid`
+based on the given level array `lvl_array`.
+The level array is has non-negative integer values.
+Value 0 means that the corresponding cell of the coarse `grid` is not subdivided
+any further. Value `n > 0` means that the corresponding cell will be subdivided
+in total `2^n` times (along each dimension), resulting in finer cells
+within the original coarse cell.
```
```suggestion
Given a coarse `grid` tesselating the state space, construct a `SubdivisionBasedGrid`
based on the given level array `lvl_array` that should have the same dimension as `grid`. 
The level array has non-negative integer values, with 0 meaning that the 
corresponding cell of the coarse `grid` should not be subdivided any further. 
Value `n > 0` means that the corresponding cell will be subdivided
in total `2^n` times (along each dimension), resulting in finer cells
within the original coarse cell.
```

awage replied:
I am not sure about the dimension of the grid. You can correct this.  

Another point: this is the documentation of a type. It should be explicit in the docstring?

awage commented on src/mapping/recurrences/grids.jl:
```diff
@@ -0,0 +1,169 @@
+        end
+    end
+
+    maxvel = maximum(filter(x -> x != Inf, velocities))
+    velratios = maxvel./velocities
+    result = [round(Int,log2(clamp(x, 1, 2^maxlevel))) for x in velratios]
```
This procedure is documented somewhere? What is the criterion for the subdivision? Why the normalization of the velocities is upside down? I would have expected `velratios = velocities/mavel`

awage commented:
I am done with the review. There is the function `update_fsm_state!` that should be changed. I will think about a better way to write this function.

Datseris commented:
Thanks @awage. I have addressed all comments. Please leave an _approving_ review (by clicking review, and then selecting approve), if you think this should be merged.
